dmk2jv3.man
.TH dmk2jv3 1
.SH Name
dmk2jv3 \- Convert a DMK floppy disk image file to JV3 format
.SH Syntax
.B dmk2jv3 [options] filename.dmk [filename.dsk]
.SH Description
The dmk2jv3 program converts a floppy disk image in the DMK format to
the JV3 format.  Both these formats were originally developed for
TRS-80 Model I/III/4/4P emulators, though the DMK format is useful
more broadly.  This converter is needed because some older emulators
and tools do not support DMK but do support JV3.

The DMK format is more general than JV3; it can represent error
conditions and other properties of the original disks that are not
representable in the JV3 format.  The converter prints an error
message whenever an important property of the DMK image cannot be
represented in the JV3 image, but when possible, it continues the
conversion.  Some less important properties of the DMK format are
silently lost, such as the exact width and content of inter-sector
gaps.

A few features of the JV3 format that were added as extensions to the
original definition are not supported by some emulators.  The
converter uses these extensions when needed, but prints a warning
message.

For more information about
the DMK and JV3 file formats and the emulators that use them, see:
.nf
    http://discover-net.net/~dmkeil/
    http://www.tim-mann.org/xtrs.html
.fi
.SH Options
If only filename.dmk is given on the command line, filename.dsk is
formed by deleting any extension from filename.dmk and appending ".dsk".
.TP
.B \-v \fIverbosity\fP
Specify how much output is printed.  Larger numbers select more output.
The default setting is 1.
.RS
.TP
0
No output except for error messages.
.TP
1
Also print warning messages.
.TP
2
Also print the sector numbers found.  Print a "?" after each sector
recorded with a CRC error.
.SH Diagnostics
.TP
.B dmk2jv3: Error reading from DMK file
The DMK file could not be read.
.TP
.B dmk2jv3: File is not in DMK format
The specified input file was not in DMK format.
.TP
.B dmk2jv3: Error writing to JV3 file
The JV3 file could not be written.
.TP
.B JV3 does not support RX02 encoding; faking it
The specified input file uses DEC RX02 encoding.  The JV3 file format
does not support RX02 encoding, so the JV3 file is written as if the
disk were in IBM standard encoding.  RX02 double density sectors are
written as IBM standard MFM, with DAM 0xf9 (RX02 deleted double
density data) changed to 0xf8 (standard MFM deleted double density
data), and DAM 0xfd (RX02 double density data) changed to 0xfb
(standard MFM double density data).  Any single density sectors
present (DAM 0xf8 or 0xfb) are represented unchanged, as DEC RX02 uses
IBM standard FM encoding for such sectors.
.TP
.B Too many total sectors
The extended JV3 format has a limit of 5802 total sectors.
Any sectors beyond this are not written to the JV3 file.
.TP
.B Warning: too many total sectors for some emulators
The original JV3 format has a limit of 2901 total sectors.
Sectors beyond this point are written to the JV3 file, but some
emulators will not be able to read them.
.TP
.B Track number xx too large, using yy
The JV3 format can only record track numbers less than 255.
Track 255 is written to the JV3 file with track number 254.
.TP
.B Sector size code xx too large, using yy
The extended JV3 format can only record sector size codes 0 (128
bytes) through 3 (1024 bytes).  The JV3 is written with the
higher-order bits of the size code masked off.
.TP
.B Warning: sector size xx not supported by some emulators
The original JV3 format can only record sectors of size 256.
The sector is recorded at its correct size, but the resulting
JV3 will not work with some emulators.
.TP 
.B Single density DAM xx not supported, using yy
.PD 0
.TP
.B Double density DAM xx not supported, using yy
.PD
The JV3 format can support only a limited range of data address marks.
.TP
.B False track number xx not supported
The JV3 format requires the logical track number recorded in a
sector's id to match the true physical track number.  The true
physical track number is recorded on the JV3 in place of the false
one.  The result will probably not work as desired.  This error is
typical of many copy-protected disks.
.TP
.B Warning: False side number xx not supported, using yy
.PD 0
.TP
.B Side number xx too large, using yy
.PD
The JV3 format requires the logical side number recorded in a
sector's id to match the true physical side number (0 or 1).  The JV3 
is written with the true side number.  This is generally not a problem 
unless the false side number was part of a copy-protection scheme.
.TP
.B Recording ID CRC error as data CRC error
The JV3 format has no way to indicate that a sector ID has a CRC
error, so we record this condition as if it were a data CRC error.
.TP
.B Recording missing DAM as data CRC error
The JV3 format has no way to indicate that a sector ID has no matching
data address mark, so we record this condition as if it were a data CRC error.
.TP
.B Warning: JV3 does not support quirks; ignoring
The DMK input file was created using one or more cw2dmk quirk flags.
JV3 format does not support any quirks, so the output file may be
missing information and/or may have false CRC errors.
.SH Author
dmk2jv3 was written by Timothy Mann, http://tim-mann.org/.
dmk2jv3 is free software, released under the GNU General Public License.
Thanks to David Keil for designing and
documenting the DMK file format for floppy disk images,
and to Jeff Vavasour for designing and documenting the JV3 format.



ChangeLog
4.9.3 -- Sun Nov 27 20:04:04 UTC 2022 -- Tim Mann

* Add QUIRK_MFM_CLOCK.  This quirk was needed to read some older UDOS
  disks with clock bits set in certain places that violate the MFM
  clocking rules.  The most frequent violation seen on a sample disk
  was that in a sequence of bytes with data value ff ff ff..., the
  clocking pattern could be 00 88 88... instead of 00 00 00...  This
  results in flux transitions in adjacent bit cells, which is normally
  illegal in MFM.  The disks were probably from UDOS 1526 V.3, or
  possibly V.2.

4.9.2 -- Sun Sep 18 23:11:36 UTC 2022 -- Tim Mann

* Make the -v2 and -v3 output of dmk2cw more readable.

* Check that numeric option arguments are valid numbers.

* Make dmk2cw warn if not all sector data fits on the physical track.
  To help support the feature (and because it generally seems useful),
  cw2dmk now records the quirk flags in the previously unused/reserved
  byte 5 of the DMK file header.

4.9.1 -- Mon May 23 15:56:42 UTC 2022 -- Tim Mann

* Add -T step_time,settling_time feature to both cw2dmk and dmk2cw.

* Log all Catweasel samples when at verbosity level 7, instead of
  stopping when the DMK buffer is full.  (This change was actually
  made in 4.9.0, but it had a bug that is now fixed.)

4.9.0 -- Sat Apr 30 17:07:57 UTC 2022 -- Tim Mann, Quentin Barnes

* Catch errors when writing to files.   Report and exit when errors
  are seen.

* Add -X min_retry feature to cw2dmk, accepting the same syntax for
  min_retry as for -x max_retry.

* Add QUIRK_IAM to -q option in cw2dmk.

* Various minor cleanup of code and the cw2dmk usage message.

* Add -R (replay) feature to cw2dmk.

* Add QUIRK_EXTRA_DATA to -q option in cw2dmk.

4.8.1 -- Sat Dec 11 01:12:14 UTC 2021 -- Quentin Barnes

* Correct a substantial performance regression when doing high
  level writes to the log file.

4.8.0 -- Tue Dec  7 20:52:50 UTC 2021 -- Quentin Barnes

* This release is virtually all of Tim Mann's work, and most of
  those changes are for adding UDOS support.

* Add -q (quirk) option for extending cw2dmk to better support of
  reading UDOS formatted disks.  UDOS disks may have a modified
  ID premark and after each data sector may include an extra 4 data
  bytes and a possible 2 bytes of CRC.  For details, see the cw2dmk
  man page's description of the -q option.

  The new -q option hasn't been tested with the -j option.  Bugs
  may show up.  If found, please report them.

* Add -a rate_adj and -e dither options to dmk2cw.

4.7.1 -- Wed Oct 20 20:00:21 UTC 2021 -- Quentin Barnes

* Add feature to "-x" option to allow per-track retry counts.
  (E.g. "-x 40,0:4-12" sets retries for all tracks to 40 except 
  disable retries for tracks 4 through 12.)  See cw2dmk man
  page for further information.

* Correct off-by-one problem with retry counts.

4.7.0 -- Mon Apr 12 23:25:25 UTC 2021 -- Quentin Barnes

* Deprecate cw2dmk's -b option introduced in 4.6.0.  It had too
  many shortcomings with CP/M disks.  Replaced with -C.

* Add -C option to cw2dmk for controlling detection of incompatibly
  formatted sides of a floppy disk.  If side 1 is formatted with
  512-byte sectors and side 0 is formatted with any other sector
  size, it'll assume side 1 was factory-formatted and skip it.
  Detection of other incompatible combinations may be added in a
  future release.  Use "-C 0" to disable the detection.

* Add new interactive menu.  When enabled with -Mi and the interrupt
  key is pressed (typically ^C), cw2dmk will enter a menu with
  options.  Also, -Me will enable dropping into the menu when the
  error limit for a track exceeds the number set by -x.

4.6.0 -- Thu Dec 17 17:24:34 UTC 2020 -- Quentin Barnes

* Skipping over 4.5.x versions since George Phillips had used that
  point release line for his work.

* Git repo "cw2dmk" of Tim Mann's work was created by ezrec
  (Jason S. McMullan).  Transferred to qbarnes.

* Switched Linux release binaries to be for 64-bit OSes.

* Added change from George Phillips for his "join" (-j) feature
  that will join different rereads of the same track and merge
  good sectors.

* Fixed problem of missing data between index hole detect and read
  started.  This should noticably reduce disk read errors and even
  occasionally dropping of entire sectors.

* Added change to allow cw2dmk tools to function on newer Linux
  releases by enabling PCI I/O space access if disabled.

* Added change to recognize and skip factory-formatted sides of a
  disk.  Controlled with -b option.

* The subdirectory "cwsdpmi" is now a git sub-module for picking up
  the "cwsdpmi.exe" binary for MS-DOS.

* Cleaned up and merged multiple makefiles for different environments
  into one.  Makes it easier to maintain, build, and cross-build.  By
  default, the Makefile will now only build for the native OS.  See
  the new file BUILDING.md for further information.

* Created subdirectory "firmware" for holding firmware related files
  including a LICENSE file.  I also received some additional
  documentation from Jens Schoenfeld on programming the MK3 and MK4
  and a file with firmware revision history saved here as
  firmware/Changelog.

* To be less generic, renamed "testhist" binary to "cwtsthst", still
  keeping to MS-DOS's 8.3 file naming convention.

* Update code base to be 64-bit clean.  Fixed all known problems but
  some may remain.

* Changed to using semantic versioning for version ID.

4.4 -- Sat Nov 28 11:45:01 PST 2009 -- Tim Mann

* The Linux binaries included in this distribution are compiled on a
  32-bit Ubuntu 8.04.3 LTS system.  The MS-DOS binaries are still
  compiled with a version of djgpp from the year 2000, under djgpp.
  The MS-DOS binaries are untested at the time of this release.
	
* Updated to the current Catweasel MK4 firmware from Individual
  Computers, moving from pre29.cw4 to rel2f2.cw4.  Changed catweasl.c
  slightly to work with the new firmware -- the write pulse width now
  must be set.


dmk2jv3.c
/*
 * dmk2jv3: Convert a DMK format emulated floppy to JV3 format if possible
 * Copyright (C) 2002 Timothy Mann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <math.h>
#include <errno.h>

#include "dmk.h"
#include "jv3.h"
#include "crc.c"

/* Command-line parameters */
#define OUT_MIN 0
#define OUT_QUIET 0
#define OUT_WARNINGS 1
#define OUT_SECTORS 2
#define OUT_MAX 2
int out_fmt = OUT_WARNINGS;

void usage()
{
  printf("\nUsage: dmk2jv3 [options] file.dmk [file.dsk]\n");
  printf(" Options [defaults in brackets]:\n");
  printf(" -v verbosity  Amount of output [%d]\n", out_fmt);
  printf("               0 = Print only errors\n");
  printf("               1 = Also print warnings\n");
  printf("               2 = Also print sector numbers found\n");
  exit(1);
}

void
dmk_read_header(FILE* dmk_file, dmk_header_t* dmk_header)
{
  int ret;
  rewind(dmk_file);
  /* assumes this machine is little-endian: */
  ret = fread(dmk_header, sizeof(dmk_header_t), 1, dmk_file);
  if (ret != 1) {
    fprintf(stderr, "dmk2jv3: Error reading from DMK file\n");
    perror("dmk2jv3");
    exit(1);
  }
}

void
dmk_inc(int *datap, int tracklen, int density, int fmtimes)
{
  *datap = *datap + (density ? 1 : fmtimes);
  if (*datap > tracklen) *datap = DMK_TKHDR_SIZE;
}

/* State for JV3 being written */
typedef struct {
  unsigned char track;
  unsigned char sector;
  unsigned char flags;
} SectorId;

typedef struct {
  int next;
  SectorId id[JV3_SECSMAX+1];
  int errcount;
  int warncount;
  long id2addr;
  int warnSize, warnIDCRC, warnMissingDAM;
} JV3State;

static JV3State jv3;

void
jv3_reset()
{
  jv3.next = 0;
  memset(jv3.id, 0xff, sizeof(jv3.id));
  jv3.errcount = 0;
  jv3.warncount = 0;
  jv3.id2addr = 0;
  jv3.warnSize = 0;
}

void
jv3_id(unsigned char track, unsigned char side,
       unsigned char sector, unsigned char sizecode,
       unsigned char dam, int density, int crcerror)
{
  SectorId id;

  if (jv3.next == JV3_SECSMAX) {
    printf("[Too many total sectors (%d)]\n", jv3.next + 1);
    jv3.errcount++;
    jv3.next++;
    return;
  }
  if (jv3.next == JV3_SECSMAX/2 && out_fmt >= OUT_WARNINGS) {
    printf("[Warning: too many total sectors for some emulators (> %d)]\n",
	   jv3.next);
    jv3.warncount++;
  }

  if (track >= JV3_TRACKSMAX) {
    printf("[Track number %u too large, using %u]\n", track, JV3_TRACKSMAX-1);
    jv3.errcount++;
    track = JV3_TRACKSMAX-1;
  }
  id.track = track;

  if (side > 1) {
    /* Can't really happen; main passes the physical side number here. */
    printf("[Side number 0x%02x too large, using %u]\n", side, side & 1);
    jv3.errcount++;
    side = side & 1;
  }
  id.flags = side ? JV3_SIDE : 0;

  id.sector = sector;

  if (sizecode > 3) {
    printf("[Sector size code 0x%02x too large, using %d]\n",
	   sizecode, sizecode & 3);
    jv3.errcount++;
    sizecode = sizecode & 3;
  }
  if (sizecode != 1 && out_fmt >= OUT_WARNINGS && !jv3.warnSize) {
    printf("[Warning: sector size %d not supported by some emulators]\n",
	   128 << sizecode);
    jv3.warncount++;
    jv3.warnSize = 1;
  }
  id.flags |= (sizecode ^ 1);

  if (density == 0) {
    /* Single density */
    switch (dam) {
    case 0xf8:
      id.flags |= JV3_DAMSDF8;
      break;
    case 0xf9:
      id.flags |= JV3_DAMSDF9;
      break;
    case 0xfa:
      id.flags |= JV3_DAMSDFA;
      break;
    case 0xfb:
      id.flags |= JV3_DAMSDFB;
      break;
    default:
      printf("[Single density DAM 0x%02x not supported, using 0xf8]\n", dam);
      jv3.errcount++;
      id.flags |= JV3_DAMSDF8;
      break;
    }
  } else {
    /* Double density */
    id.flags |= JV3_DENSITY;
    switch (dam) {
    case 0xf8:
    case 0xf9: // RX02 fake support; record as deleted MFM data
      id.flags |= JV3_DAMDDF8;
      break;
    case 0xfb:
    case 0xfd: // RX02 fake support; record as normal MFM data
      id.flags |= JV3_DAMDDFB;
      break;
    default:
      printf("[Double density DAM 0x%02x not supported, using 0xf8]\n", dam);
      jv3.errcount++;
      id.flags |= JV3_DAMDDF8;
      break;
    }
  }

  if (crcerror) id.flags |= JV3_ERROR;

  jv3.id[jv3.next++] = id;
}

int
main(int argc, char** argv)
{
  char* dmk_name = NULL;
  char* jv3_name = NULL;
  FILE* dmk_file;
  FILE* jv3_file;
  dmk_header_t dmk_header;
  unsigned char* dmk_track;
  int ch, ret;
  int track, side, sides, fmtimes, rx02;
  int idampp, idamp, datap;
  int density;
  int dam_min, dam_range;

  opterr = 0;
  for (;;) {
    ch = getopt(argc, argv, "v:");
    if (ch == -1) break;
    switch (ch) {
    case 'v':
      out_fmt = strtol(optarg, NULL, 0);
      if (out_fmt < OUT_MIN || out_fmt > OUT_MAX) usage();
      break;
    default:
      usage();
      break;
    }
  }

  switch (argc - optind) {
  case 2:
    /* Two filenames given */
    dmk_name = argv[optind];
    jv3_name = argv[optind+1];
    break;

  case 1: {
    char *p;
    int len;

    dmk_name = argv[optind];
    p = strrchr(dmk_name, '.');
    if (p == NULL) {
      len = strlen(dmk_name);
    } else {
      len = p - dmk_name;
    }
    jv3_name = (char *) malloc(len + 5);
    sprintf(jv3_name, "%.*s.dsk", len, dmk_name);
    break; }

  default:
    usage();
  }

  dmk_file = fopen(dmk_name, "rb");
  if (dmk_file == NULL) {
    perror(dmk_name);
    exit(1);
  }

  /* Set DMK parameters */
  dmk_read_header(dmk_file, &dmk_header);
  if ((dmk_header.writeprot != 0x00 && dmk_header.writeprot != 0xff) ||
      dmk_header.mbz != 0) {
    fprintf(stderr, "dmk2jv3: File is not in DMK format\n");
    exit(1);
  }
  sides = (dmk_header.options & DMK_SSIDE_OPT) ? 1 : 2;
  fmtimes = (dmk_header.options & DMK_SDEN_OPT) ? 1 : 2;
  rx02 = (dmk_header.options & DMK_RX02_OPT) ? 1 : 0;
  dmk_track = (unsigned char*) malloc(dmk_header.tracklen);
  if (out_fmt >= OUT_SECTORS) {
    printf("[tracks=%d, sides=%d, fmtimes=%d, rx02=%d, tracklen=0x%x]\n",
	   dmk_header.ntracks, sides, fmtimes, rx02, dmk_header.tracklen);
  }

  if (rx02) {
    printf("[JV3 does not support RX02 encoding; faking it]\n");
  }
  if (dmk_header.quirks != 0 && out_fmt >= OUT_WARNINGS) {
    printf("[Warning: JV3 does not support quirks; ignoring %02x]\n",
           dmk_header.quirks);
    jv3.warncount++;
  }

  jv3_file = fopen(jv3_name, "wb");
  if (jv3_file == NULL) {
    perror(jv3_name);
    exit(1);
  }
  fseek(jv3_file, JV3_SECSTART, 0);
  jv3_reset();

  /* Loop through tracks */
  for (track=0; track<dmk_header.ntracks; track++) {

    /* Loop through sides */
    for (side=0; side<sides; side++) {
      /* Read DMK track data */
      ret = fread(dmk_track, dmk_header.tracklen, 1, dmk_file);
      if (ret != 1) {
	if (errno == 0) {
	  printf("[End of file on input]\n");
	  break;
	}
	fprintf(stderr, "dmk2jv3: Error reading from DMK file\n");
	perror("dmk2jv3");
	exit(1);
      }

      if (out_fmt >= OUT_SECTORS) {
	printf("Track %d, side %d, sector: ", track, side);
      }

      /* Loop through ids */
      idampp = 0;
      for (;;) {
	idamp = dmk_track[idampp++];
	idamp += dmk_track[idampp++] << 8;
	if (idamp == 0 || idamp == 0xffff) break;
	density = (idamp & DMK_DDEN_FLAG) != 0;
	idamp &= DMK_IDAMP_BITS;

	/* Project where DAM will be */
	if (!density) {
	  dam_min = idamp + 7 * fmtimes;
	  dam_range = 30 * fmtimes;  /* ref 1791 datasheet */
	} else {
	  dam_min = idamp + 7;
	  dam_range = 43;  /* ref 1791 datasheet */
	}

	/* Decode an id block if idamp is valid */
	datap = idamp;
	if (dmk_track[datap] == 0xfe) {
	  unsigned char ltrack, lside, sector, sizecode, dam = 0;
	  int crcerror = 0, size;
	  unsigned short crc;
#         define DMK_INC(d) dmk_inc(&d, dmk_header.tracklen, density, fmtimes)

	  /* Start ID CRC check */
	  if (density == 0) {
	    crc = 0xffff;
	  } else {
	    crc = calc_crc1(0x968b, 0xa1); /* CRC of a1 a1 a1 */
	  }
	  crc = calc_crc1(crc, 0xfe);
	  DMK_INC(datap);
	  ltrack = dmk_track[datap];
	  crc = calc_crc1(crc, ltrack);
	  DMK_INC(datap);
	  lside = dmk_track[datap];
	  crc = calc_crc1(crc, lside);
	  DMK_INC(datap);
	  sector = dmk_track[datap];
	  crc = calc_crc1(crc, sector);
	  DMK_INC(datap);
	  sizecode = dmk_track[datap];
	  crc = calc_crc1(crc, sizecode);
	  size = 128 << (sizecode & 3);

	  if (out_fmt >= OUT_SECTORS) {
	    printf(" %d", sector);
	  }

	  if (ltrack != track) {
	    printf("[False track number 0x%02x not supported]\n", ltrack);
	    jv3.errcount++;
	  }

	  if (lside != side && out_fmt >= OUT_WARNINGS) {
	    printf("[Warning: False side number 0x%02x not supported, "
		   "using %d]\n", lside, side);
	    jv3.warncount++;
	  }

	  /* Check ID CRC */
	  DMK_INC(datap);
	  crc = calc_crc1(crc, dmk_track[datap]);
	  DMK_INC(datap);
	  crc = calc_crc1(crc, dmk_track[datap]);
	  if (crc != 0) {
	    printf("[Recording ID CRC error as data CRC error]\n");
	    jv3.errcount++;
	    crcerror = 1;
	  }

	  /* Look for dam */
	  datap = dam_min;
	  while (--dam_range >= 0) {
	    dam = dmk_track[datap];
	    DMK_INC(datap);
	    if ((dam >= 0xf8 && dam <= 0xfb) ||
                (dam == 0xfd && rx02)) break;
	  }
	  if (dam_range < 0) {
	    printf("[Recording missing DAM as data CRC error]\n");
	    jv3.errcount++;
	    dam = 0xfb;
	  }

	  /* Start data CRC check */
	  if (density == 0) {
	    crc = 0xffff;
	  } else {
	    crc = calc_crc1(0x968b, 0xa1); /* CRC of a1 a1 a1 */
	  }
	  crc = calc_crc1(crc, dam);

          /* RX02 double density data?  Fake it as standard MFM. */
          if (rx02 && (dam == 0xf9 || dam == 0xfd)) {
            sizecode += 1;
            size <<= 1;
            density = 1;
          }

	  /* Write out data */
	  while (size--) {
	    putc(dmk_track[datap], jv3_file);
	    crc = calc_crc1(crc, dmk_track[datap]);
	    DMK_INC(datap);
	  }

	  /* Check data CRC */
	  crc = calc_crc1(crc, dmk_track[datap]);
	  DMK_INC(datap);
	  crc = calc_crc1(crc, dmk_track[datap]);
	  if (crc != 0) {
	    crcerror = 1;
	  }

	  if (out_fmt >= OUT_SECTORS && crcerror) {
	    printf("?");
	  }

	  /* Remember id info */
	  jv3_id(track, side, sector, sizecode, dam, density, crcerror);

	  /* Leave space for second id block if needed */
	  if (jv3.next == JV3_SECSPERBLK) {
	    jv3.id2addr = ftell(jv3_file);
	    fseek(jv3_file, jv3.id2addr + JV3_SECSTART, 0);
	  }
	}
      }

      if (out_fmt >= OUT_SECTORS) {
	printf("\n");
	fflush(stdout);
      }
    }
  }

  /* Write out id blocks and writeprot flag */
  fseek(jv3_file, 0, 0);
  ret = fwrite(jv3.id, JV3_SECSTART - 1, 1, jv3_file);
  if (ret != 1) {
    fprintf(stderr, "dmk2jv3: Error writing to JV3 file\n");
    perror("dmk2jv3");
    exit(1);
  }
  fputc(0xff, jv3_file);
  if (jv3.id2addr) {
    fseek(jv3_file, jv3.id2addr, 0);
    fwrite(((unsigned char*) jv3.id) + JV3_SECSTART - 1,
	   JV3_SECSTART, 1, jv3_file);
  }
  if (out_fmt > OUT_QUIET || jv3.errcount > 0 || jv3.warncount > 0) {
    printf("%d total errors, %d total warnings\n",
	   jv3.errcount, jv3.warncount);
  }
  return 0;
}
